File: youbot_scanner_viz/src/scannerViz.py
#!/usr/bin/env python
"""
This script shows the distances of the primary directions of the LaserScanner.
"""

from os import system
import rospy
from sensor_msgs.msg import LaserScan
from math import pi, sin, cos

### BEGIN MAGIC NUMBERS
NROFDIRS      = 9
# if this is changed, printMainDirections has to be changed as well!
NROFDIGITS    = 3
SCANNERANGLE  = 180
NROFDATAPOINTS = 640
### END MAGIC NUMBERS


# execute this every time data is received
def laserCallback(data):
  seqNr = data.header.seq
  laser = laserCalc(data)
  laser = snipDirections(laser)
  printMainDirections(seqNr, laser[0], laser[1])

"""
Print directions in a nice way.
"""
def printMainDirections(seqNr, nrOfDirs, ranges):
  angles = []
  for dir in range(nrOfDirs):
    angles.append(dir * SCANNERANGLE / (nrOfDirs-1))
  # prettyPrint - Build the Strings
  lineString = "+"
  for column in range(nrOfDirs):
    lineString += "-------+"
  angleString = "|"
  for angle in angles:
    angleString += "  %3.0f  |" % angle
  rangeString = "|"
  for rangei in ranges:
    rangeString += " %1.3f |" % rangei
  # prettyPrint - Print the Strings
  print("Sequence Number: %10u" % seqNr)
  print(lineString)
  print(angleString)
  print(lineString)
  print(rangeString)
  print(lineString)
  print("\n")

"""
Convert the raw data from the laserScan to an array of distances.
"""
def laserCalc(data):
  laser = []
  for range in data.ranges:
    laser.append(range)
  return laser

"""
Take only "main" directions from the whole 640 data points.
"""
def snipDirections(ranges):
  mainRanges = []
  for i in range(NROFDIRS):
    if (i != 0):
      mainRanges.append(round(ranges[(i*NROFDATAPOINTS/(NROFDIRS-1))-1],NROFDIGITS))
    else:
      mainRanges.append(round(ranges[0], NROFDIGITS))
  mainRanges[0] = round(ranges[0], NROFDIGITS)
  return (NROFDIRS, mainRanges)

def init():
  # Possibly mutliple clients want to display scanner data
  rospy.init_node('scannerViz', anonymous=True)
  # Subscribe to Sensor data
  rospy.Subscriber("/youbot/scan_front", LaserScan, laserCallback)
  # Don't close this until node is stopped
  rospy.spin()



if __name__ == '__main__':
  init()

File: youbot_scanner_viz/src/scannerViz.cpp
#include "ros/ros.h"
#include "sensor_msgs/LaserScan.h"

#include <iostream>
using namespace std;

/**
 * This program shows the distances of the primary directions of the LaserScanner.
 */

// Begin Magic Numbers
const uint NROFDIRS       = 9;
// if this is changed, printMainDirections has to be changed as well!
const uint NROFDIGITS     = 3;
const uint SCANNERANGLE   = 180;
const uint NROFDATAPOINTS = 640;
// End Magic Numbers

/**
 * Is called upon receiveing a LaserScan message.
 */
void laserCallback(const sensor_msgs::LaserScan::ConstPtr& data)
{
  const uint seqNr            = data->header.seq;
  const uint nrOfMeasurements = sizeof(data->ranges)/sizeof(data->ranges[0]);

  // put range values in array
  float laserArray[NROFDATAPOINTS] = {};
  for (uint i=0; i<NROFDATAPOINTS; i++)
  {
    laserArray[i] = data->ranges[i];
  }

  // take only main directions
  float directions[NROFDIRS] = {};
  for (uint i=0; i<NROFDIRS; i++)
  {
    if (i != 0)
      directions[i] = laserArray[(int) (i*NROFDATAPOINTS/(NROFDIRS-1))-1];
    else
      directions[i] = laserArray[0];
  }

  // print out values
  // prepare separator
  string lineString      = "+";
  for (uint i=0; i<NROFDIRS; i++)
  {
    lineString  += "-------+";
  }
  lineString += "\n";

  printf("\nSequence Number: %10u", seqNr);
//   cout << "\nSequence Number: " << seqNr << "\n";
  cout << "\n" << lineString;
  // header - angles
  cout << "|";
  for (uint i=0; i<NROFDIRS; i++)
  {
    printf(" %5u |", i*SCANNERANGLE / (NROFDIRS-1));
  }
  cout << "\n";
  cout << lineString;
  // body - ranges
  cout << "|";
  for (uint i=0; i<NROFDIRS; i++)
  {
    printf(" %1.3f |", directions[i]);
  }
  cout << "\n";
  cout << lineString;
}

int main(int argc, char **argv)
{
  // Initialize ROS system
  ros::init(argc, argv, "scanner_viz");

  // Initialize node
  ros::NodeHandle n;

  // Subscribe to the LaserScan topic
  ros::Subscriber laserSub = n.subscribe("/youbot/scan_front", 5, laserCallback);

  // keep alive
  ros::spin();

  return 0;
}
